From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tom Sepez <tsepez@google.com>
Date: Wed, 14 Oct 2026 11:02:52 -0700
Subject: CHECK on FFE teardown if embedder is inside a callback.

This indicates a misbehaving embedder, so flag it with a CHECK()
while it is still active on the stack.

-- mass rename two hungarian length variables.
-- consistently name AutoRestorer local `restorer`

Bug: 572237007
Change-Id: Ice9212769e4072d76a427379aa13d02fa36be55d
Reviewed-on: https://pdfium-review.googlesource.com/c/pdfium/+/160615
Reviewed-by: Lei Zhang <thestig@chromium.org>
Commit-Queue: Tom Sepez <tsepez@chromium.org>

diff --git a/fpdfsdk/cpdfsdk_formfillenvironment.cpp b/fpdfsdk/cpdfsdk_formfillenvironment.cpp
index 1568fc975b0be15b0bb812cb57fcd2d0df3d5101..5f609c421316934f79613c689ac329a6e85e3dbe 100644
--- a/fpdfsdk/cpdfsdk_formfillenvironment.cpp
+++ b/fpdfsdk/cpdfsdk_formfillenvironment.cpp
@@ -74,6 +74,8 @@ CPDFSDK_FormFillEnvironment::CPDFSDK_FormFillEnvironment(
 }
 
 CPDFSDK_FormFillEnvironment::~CPDFSDK_FormFillEnvironment() {
+  CHECK(!in_active_ffi_callback_);
+  CHECK(!in_passive_ffi_callback_);
   being_destroyed_ = true;
   ClearAllFocusedAnnots();
 
@@ -122,7 +124,7 @@ void CPDFSDK_FormFillEnvironment::OutputSelectedRect(
   CFX_PointF ptA = pFormField->PWLtoFFL(CFX_PointF(rect.left, rect.bottom));
   CFX_PointF ptB = pFormField->PWLtoFFL(CFX_PointF(rect.right, rect.top));
 
-  AutoRestorer<bool> ffi_restorer(&in_passive_ffi_callback_);
+  AutoRestorer<bool> restorer(&in_passive_ffi_callback_);
   in_passive_ffi_callback_ = true;
   info_->FFI_OutputSelectedRect(info_, pPage, ptA.x, ptB.y, ptB.x, ptA.y);
 }
@@ -137,8 +139,10 @@ CPDFSDK_PageView* CPDFSDK_FormFillEnvironment::GetCurrentView() {
   return pPage ? GetOrCreatePageView(pPage) : nullptr;
 }
 
-IPDF_Page* CPDFSDK_FormFillEnvironment::GetCurrentPage() const {
+IPDF_Page* CPDFSDK_FormFillEnvironment::GetCurrentPage() {
   if (info_ && info_->FFI_GetCurrentPage) {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
     return IPDFPageFromFPDFPage(info_->FFI_GetCurrentPage(
         info_, FPDFDocumentFromCPDFDocument(cpdfdoc_)));
   }
@@ -151,19 +155,29 @@ WideString CPDFSDK_FormFillEnvironment::GetLanguage() {
     return WideString();
   }
 
-  int nRequiredLen = info_->FFI_GetLanguage(info_, nullptr, 0);
-  if (nRequiredLen <= 0) {
+  int required_len;
+  {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
+    required_len = info_->FFI_GetLanguage(info_, nullptr, 0);
+  }
+  if (required_len <= 0) {
     return WideString();
   }
 
-  DataVector<uint8_t> pBuff(nRequiredLen);
-  int nActualLen = info_->FFI_GetLanguage(info_, pBuff.data(), nRequiredLen);
-  if (nActualLen <= 0 || nActualLen > nRequiredLen) {
+  DataVector<uint8_t> pBuff(required_len);
+  int actual_len;
+  {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
+    actual_len = info_->FFI_GetLanguage(info_, pBuff.data(), required_len);
+  }
+  if (actual_len <= 0 || actual_len > required_len) {
     return WideString();
   }
 
   return WideString::FromUTF16LE(
-      pdfium::span(pBuff).first(static_cast<size_t>(nActualLen)));
+      pdfium::span(pBuff).first(static_cast<size_t>(actual_len)));
 #else   // PDF_ENABLE_XFA
   return WideString();
 #endif  // PDF_ENABLE_XFA
@@ -175,19 +189,29 @@ WideString CPDFSDK_FormFillEnvironment::GetPlatform() {
     return WideString();
   }
 
-  int nRequiredLen = info_->FFI_GetPlatform(info_, nullptr, 0);
-  if (nRequiredLen <= 0) {
+  int required_len;
+  {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
+    required_len = info_->FFI_GetPlatform(info_, nullptr, 0);
+  }
+  if (required_len <= 0) {
     return WideString();
   }
 
-  DataVector<uint8_t> pBuff(nRequiredLen);
-  int nActualLen = info_->FFI_GetPlatform(info_, pBuff.data(), nRequiredLen);
-  if (nActualLen <= 0 || nActualLen > nRequiredLen) {
+  DataVector<uint8_t> pBuff(required_len);
+  int actual_len;
+  {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
+    actual_len = info_->FFI_GetPlatform(info_, pBuff.data(), required_len);
+  }
+  if (actual_len <= 0 || actual_len > required_len) {
     return WideString();
   }
 
   return WideString::FromUTF16LE(
-      pdfium::span(pBuff).first(static_cast<size_t>(nActualLen)));
+      pdfium::span(pBuff).first(static_cast<size_t>(actual_len)));
 #else   // PDF_ENABLE_XFA
   return WideString();
 #endif  // PDF_ENABLE_XFA
@@ -245,20 +269,20 @@ WideString CPDFSDK_FormFillEnvironment::JS_fieldBrowse() {
     return WideString();
   }
 
-  const int nRequiredLen = js_platform->Field_browse(js_platform, nullptr, 0);
-  if (nRequiredLen <= 0) {
+  const int required_len = js_platform->Field_browse(js_platform, nullptr, 0);
+  if (required_len <= 0) {
     return WideString();
   }
 
-  DataVector<uint8_t> pBuff(nRequiredLen);
-  const int nActualLen =
-      js_platform->Field_browse(js_platform, pBuff.data(), nRequiredLen);
-  if (nActualLen <= 0 || nActualLen > nRequiredLen) {
+  DataVector<uint8_t> pBuff(required_len);
+  const int actual_len =
+      js_platform->Field_browse(js_platform, pBuff.data(), required_len);
+  if (actual_len <= 0 || actual_len > required_len) {
     return WideString();
   }
 
   // Don't include trailing NUL.
-  pBuff.resize(nActualLen - 1);
+  pBuff.resize(actual_len - 1);
 
   // Use FromDefANSI() per "local encoding" comment in fpdf_formfill.h.
   return WideString::FromDefANSI(ByteStringView(pBuff));
@@ -326,21 +350,21 @@ WideString CPDFSDK_FormFillEnvironment::GetFilePath() const {
     return WideString();
   }
 
-  const int nRequiredLen =
+  const int required_len =
       js_platform->Doc_getFilePath(js_platform, nullptr, 0);
-  if (nRequiredLen <= 0) {
+  if (required_len <= 0) {
     return WideString();
   }
 
-  DataVector<uint8_t> pBuff(nRequiredLen);
-  const int nActualLen =
-      js_platform->Doc_getFilePath(js_platform, pBuff.data(), nRequiredLen);
-  if (nActualLen <= 0 || nActualLen > nRequiredLen) {
+  DataVector<uint8_t> pBuff(required_len);
+  const int actual_len =
+      js_platform->Doc_getFilePath(js_platform, pBuff.data(), required_len);
+  if (actual_len <= 0 || actual_len > required_len) {
     return WideString();
   }
 
   // Don't include trailing NUL.
-  pBuff.resize(nActualLen - 1);
+  pBuff.resize(actual_len - 1);
 
   // Use FromDefANSI() per "local encoding" comment in fpdf_formfill.h.
   return WideString::FromDefANSI(ByteStringView(pBuff));
@@ -370,7 +394,7 @@ IJS_Runtime* CPDFSDK_FormFillEnvironment::GetIJSRuntime() {
 void CPDFSDK_FormFillEnvironment::Invalidate(IPDF_Page* page,
                                              const FX_RECT& rect) {
   if (info_ && info_->FFI_Invalidate) {
-    AutoRestorer<bool> ffi_restorer(&in_passive_ffi_callback_);
+    AutoRestorer<bool> restorer(&in_passive_ffi_callback_);
     in_passive_ffi_callback_ = true;
     info_->FFI_Invalidate(info_, FPDFPageFromIPDFPage(page), rect.left,
                           rect.top, rect.right, rect.bottom);
@@ -380,7 +404,7 @@ void CPDFSDK_FormFillEnvironment::Invalidate(IPDF_Page* page,
 void CPDFSDK_FormFillEnvironment::SetCursor(
     IPWL_FillerNotify::CursorStyle nCursorType) {
   if (info_ && info_->FFI_SetCursor) {
-    AutoRestorer<bool> ffi_restorer(&in_passive_ffi_callback_);
+    AutoRestorer<bool> restorer(&in_passive_ffi_callback_);
     in_passive_ffi_callback_ = true;
     info_->FFI_SetCursor(info_, static_cast<int>(nCursorType));
   }
@@ -389,6 +413,8 @@ void CPDFSDK_FormFillEnvironment::SetCursor(
 int CPDFSDK_FormFillEnvironment::SetTimer(int uElapse,
                                           TimerCallback lpTimerFunc) {
   if (info_ && info_->FFI_SetTimer) {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
     return info_->FFI_SetTimer(info_, uElapse, lpTimerFunc);
   }
   return CFX_Timer::HandlerIface::kInvalidTimerID;
@@ -396,13 +422,15 @@ int CPDFSDK_FormFillEnvironment::SetTimer(int uElapse,
 
 void CPDFSDK_FormFillEnvironment::KillTimer(int nTimerID) {
   if (info_ && info_->FFI_KillTimer) {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
     info_->FFI_KillTimer(info_, nTimerID);
   }
 }
 
 void CPDFSDK_FormFillEnvironment::OnChange() {
   if (info_ && info_->FFI_OnChange) {
-    AutoRestorer<bool> ffi_restorer(&in_passive_ffi_callback_);
+    AutoRestorer<bool> restorer(&in_passive_ffi_callback_);
     in_passive_ffi_callback_ = true;
     info_->FFI_OnChange(info_);
   }
@@ -411,6 +439,8 @@ void CPDFSDK_FormFillEnvironment::OnChange() {
 void CPDFSDK_FormFillEnvironment::ExecuteNamedAction(
     const ByteString& namedAction) {
   if (info_ && info_->FFI_ExecuteNamedAction) {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
     info_->FFI_ExecuteNamedAction(info_, namedAction.c_str());
   }
 }
@@ -426,7 +456,7 @@ void CPDFSDK_FormFillEnvironment::OnSetFieldInputFocusInternal(
     size_t nCharacters = text.GetLength();
     ByteString bsUTFText = text.ToUTF16LE();
     auto* pBuffer = reinterpret_cast<const unsigned short*>(bsUTFText.c_str());
-    AutoRestorer<bool> ffi_restorer(&in_passive_ffi_callback_);
+    AutoRestorer<bool> restorer(&in_passive_ffi_callback_);
     in_passive_ffi_callback_ = true;
     info_->FFI_SetTextFieldFocus(
         info_, pBuffer, pdfium::checked_cast<FPDF_DWORD>(nCharacters), bFocus);
@@ -463,14 +493,16 @@ void CPDFSDK_FormFillEnvironment::DoURIAction(const ByteString& bsURI,
   if (!info_) {
     return;
   }
-
   if (info_->version >= 2 && info_->FFI_DoURIActionWithKeyboardModifier) {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
     info_->FFI_DoURIActionWithKeyboardModifier(info_, bsURI.c_str(),
                                                modifiers.UncheckedValue());
     return;
   }
-
   if (info_->FFI_DoURIAction) {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
     info_->FFI_DoURIAction(info_, bsURI.c_str());
   }
 }
@@ -479,6 +511,8 @@ void CPDFSDK_FormFillEnvironment::DoGoToAction(int nPageIndex,
                                                int zoomMode,
                                                pdfium::span<float> fPosArray) {
   if (info_ && info_->FFI_DoGoToAction) {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
     info_->FFI_DoGoToAction(info_, nPageIndex, zoomMode, fPosArray.data(),
                             fxcrt::CollectionSize<int>(fPosArray));
   }
@@ -496,25 +530,29 @@ void CPDFSDK_FormFillEnvironment::DisplayCaret(IPDF_Page* page,
                                                double right,
                                                double bottom) {
   if (info_ && info_->version >= 2 && info_->FFI_DisplayCaret) {
-    AutoRestorer<bool> ffi_restorer(&in_passive_ffi_callback_);
+    AutoRestorer<bool> restorer(&in_passive_ffi_callback_);
     in_passive_ffi_callback_ = true;
     info_->FFI_DisplayCaret(info_, FPDFPageFromIPDFPage(page), bVisible, left,
                             top, right, bottom);
   }
 }
 
-int CPDFSDK_FormFillEnvironment::GetCurrentPageIndex() const {
+int CPDFSDK_FormFillEnvironment::GetCurrentPageIndex() {
   if (!info_ || info_->version < 2 || !info_->FFI_GetCurrentPageIndex) {
     return -1;
   }
-  return info_->FFI_GetCurrentPageIndex(info_,
-                                        FPDFDocumentFromCPDFDocument(cpdfdoc_));
+  AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+  in_active_ffi_callback_ = true;
+  return info_->FFI_GetCurrentPageIndex(
+      info_, FPDFDocumentFromCPDFDocument(cpdfdoc_));
 }
 
 void CPDFSDK_FormFillEnvironment::SetCurrentPage(int iCurPage) {
   if (!info_ || info_->version < 2 || !info_->FFI_SetCurrentPage) {
     return;
   }
+  AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+  in_active_ffi_callback_ = true;
   info_->FFI_SetCurrentPage(info_, FPDFDocumentFromCPDFDocument(cpdfdoc_),
                             iCurPage);
 }
@@ -525,6 +563,8 @@ void CPDFSDK_FormFillEnvironment::GotoURL(const WideString& wsURL) {
   }
 
   ByteString bsTo = wsURL.ToUTF16LE();
+  AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+  in_active_ffi_callback_ = true;
   info_->FFI_GotoURL(info_, FPDFDocumentFromCPDFDocument(cpdfdoc_),
                      AsFPDFWideString(&bsTo));
 }
@@ -539,8 +579,12 @@ FS_RECTF CPDFSDK_FormFillEnvironment::GetPageViewRect(IPDF_Page* page) {
   double top;
   double right;
   double bottom;
-  info_->FFI_GetPageViewRect(info_, FPDFPageFromIPDFPage(page), &left, &top,
-                             &right, &bottom);
+  {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
+    info_->FFI_GetPageViewRect(info_, FPDFPageFromIPDFPage(page), &left, &top,
+                               &right, &bottom);
+  }
 
   rect.left = static_cast<float>(left);
   rect.top = static_cast<float>(top);
@@ -552,9 +596,13 @@ FS_RECTF CPDFSDK_FormFillEnvironment::GetPageViewRect(IPDF_Page* page) {
 bool CPDFSDK_FormFillEnvironment::PopupMenu(IPDF_Page* page,
                                             int menuFlag,
                                             const CFX_PointF& pt) {
-  return info_ && info_->version >= 2 && info_->FFI_PopupMenu &&
-         info_->FFI_PopupMenu(info_, FPDFPageFromIPDFPage(page), nullptr,
-                              menuFlag, pt.x, pt.y);
+  if (info_ && info_->version >= 2 && info_->FFI_PopupMenu) {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
+    return info_->FFI_PopupMenu(info_, FPDFPageFromIPDFPage(page), nullptr,
+                                menuFlag, pt.x, pt.y);
+  }
+  return false;
 }
 
 void CPDFSDK_FormFillEnvironment::EmailTo(FPDF_FILEHANDLER* fileHandler,
@@ -564,6 +612,8 @@ void CPDFSDK_FormFillEnvironment::EmailTo(FPDF_FILEHANDLER* fileHandler,
                                           FPDF_WIDESTRING pBcc,
                                           FPDF_WIDESTRING pMsg) {
   if (info_ && info_->version >= 2 && info_->FFI_EmailTo) {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
     info_->FFI_EmailTo(info_, fileHandler, pTo, pSubject, pCC, pBcc, pMsg);
   }
 }
@@ -572,6 +622,8 @@ void CPDFSDK_FormFillEnvironment::UploadTo(FPDF_FILEHANDLER* fileHandler,
                                            int fileFlag,
                                            FPDF_WIDESTRING uploadTo) {
   if (info_ && info_->version >= 2 && info_->FFI_UploadTo) {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
     info_->FFI_UploadTo(info_, fileHandler, fileFlag, uploadTo);
   }
 }
@@ -580,6 +632,8 @@ FPDF_FILEHANDLER* CPDFSDK_FormFillEnvironment::OpenFile(int fileType,
                                                         FPDF_WIDESTRING wsURL,
                                                         const char* mode) {
   if (info_ && info_->version >= 2 && info_->FFI_OpenFile) {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
     return info_->FFI_OpenFile(info_, fileType, wsURL, mode);
   }
   return nullptr;
@@ -592,8 +646,13 @@ RetainPtr<IFX_SeekableReadStream> CPDFSDK_FormFillEnvironment::DownloadFromURL(
   }
 
   ByteString bstrURL = url.ToUTF16LE();
-  FPDF_FILEHANDLER* file_handler =
-      info_->FFI_DownloadFromURL(info_, AsFPDFWideString(&bstrURL));
+  FPDF_FILEHANDLER* file_handler;
+  {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
+    file_handler =
+        info_->FFI_DownloadFromURL(info_, AsFPDFWideString(&bstrURL));
+  }
   if (!file_handler) {
     return nullptr;
   }
@@ -619,10 +678,14 @@ WideString CPDFSDK_FormFillEnvironment::PostRequestURL(
 
   FPDF_BSTR response;
   FPDF_BStr_Init(&response);
-  info_->FFI_PostRequestURL(
-      info_, AsFPDFWideString(&bsURL), AsFPDFWideString(&bsData),
-      AsFPDFWideString(&bsContentType), AsFPDFWideString(&bsEncode),
-      AsFPDFWideString(&bsHeader), &response);
+  {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
+    info_->FFI_PostRequestURL(
+        info_, AsFPDFWideString(&bsURL), AsFPDFWideString(&bsData),
+        AsFPDFWideString(&bsContentType), AsFPDFWideString(&bsEncode),
+        AsFPDFWideString(&bsHeader), &response);
+  }
 
   // SAFETY: required from FFI callback.
   WideString wsRet = WideString::FromUTF16LE(UNSAFE_BUFFERS(
@@ -645,14 +708,18 @@ FPDF_BOOL CPDFSDK_FormFillEnvironment::PutRequestURL(
   ByteString bsData = wsData.ToUTF16LE();
   ByteString bsEncode = wsEncode.ToUTF16LE();
 
+  AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+  in_active_ffi_callback_ = true;
   return info_->FFI_PutRequestURL(info_, AsFPDFWideString(&bsURL),
                                   AsFPDFWideString(&bsData),
                                   AsFPDFWideString(&bsEncode));
 }
 
 void CPDFSDK_FormFillEnvironment::PageEvent(int iPageCount,
-                                            uint32_t dwEventType) const {
+                                            uint32_t dwEventType) {
   if (info_ && info_->version >= 2 && info_->FFI_PageEvent) {
+    AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+    in_active_ffi_callback_ = true;
     info_->FFI_PageEvent(info_, iPageCount, dwEventType);
   }
 }
@@ -771,10 +838,12 @@ void CPDFSDK_FormFillEnvironment::RemovePageView(IPDF_Page* pUnderlyingPage) {
   page_map_.erase(it);
 }
 
-IPDF_Page* CPDFSDK_FormFillEnvironment::GetPage(int nIndex) const {
+IPDF_Page* CPDFSDK_FormFillEnvironment::GetPage(int nIndex) {
   if (!info_ || !info_->FFI_GetPage) {
     return nullptr;
   }
+  AutoRestorer<bool> restorer(&in_active_ffi_callback_);
+  in_active_ffi_callback_ = true;
   return IPDFPageFromFPDFPage(info_->FFI_GetPage(
       info_, FPDFDocumentFromCPDFDocument(cpdfdoc_), nIndex));
 }
@@ -915,7 +984,7 @@ void CPDFSDK_FormFillEnvironment::SendOnFocusChange(
   FPDF_ANNOTATION fpdf_annot =
       FPDFAnnotationFromCPDFAnnotContext(focused_annot.get());
 
-  AutoRestorer<bool> ffi_restorer(&in_passive_ffi_callback_);
+  AutoRestorer<bool> restorer(&in_passive_ffi_callback_);
   in_passive_ffi_callback_ = true;
   info_->FFI_OnFocusChange(info_, fpdf_annot, pPageView->GetPageIndex());
 }
diff --git a/fpdfsdk/cpdfsdk_formfillenvironment.h b/fpdfsdk/cpdfsdk_formfillenvironment.h
index 74efbd8f7fe30005d0a45a8d20fdc96f3cbb3417..96c319047d08bdd6baf62bf8dbb2c2c53f8d67ca 100644
--- a/fpdfsdk/cpdfsdk_formfillenvironment.h
+++ b/fpdfsdk/cpdfsdk_formfillenvironment.h
@@ -148,7 +148,7 @@ class CPDFSDK_FormFillEnvironment final
 
 #ifdef PDF_ENABLE_V8
   CPDFSDK_PageView* GetCurrentView();
-  IPDF_Page* GetCurrentPage() const;
+  IPDF_Page* GetCurrentPage();
 
   WideString GetLanguage();
   WideString GetPlatform();
@@ -191,7 +191,7 @@ class CPDFSDK_FormFillEnvironment final
                     double top,
                     double right,
                     double bottom);
-  int GetCurrentPageIndex() const;
+  int GetCurrentPageIndex();
   void SetCurrentPage(int iCurPage);
 
   // TODO(dsinclair): This should probably change to PDFium?
@@ -222,7 +222,7 @@ class CPDFSDK_FormFillEnvironment final
                           const WideString& wsData,
                           const WideString& wsEncode);
 
-  void PageEvent(int iPageCount, uint32_t dwEventType) const;
+  void PageEvent(int iPageCount, uint32_t dwEventType);
 #endif  // PDF_ENABLE_XFA
 #endif  // PDF_ENABLE_V8
 
@@ -249,7 +249,7 @@ class CPDFSDK_FormFillEnvironment final
  private:
   using RunScriptCallback = std::function<void(IJS_EventContext* context)>;
 
-  IPDF_Page* GetPage(int nIndex) const;
+  IPDF_Page* GetPage(int nIndex);
   void OnSetFieldInputFocusInternal(const WideString& text, bool bFocus);
   void SendOnFocusChange(ObservedPtr<CPDFSDK_Annot>& pAnnot);
 
@@ -289,6 +289,7 @@ class CPDFSDK_FormFillEnvironment final
   std::unique_ptr<CFFL_InteractiveFormFiller> interactive_form_filler_;
   bool change_mask_ = false;
   bool being_destroyed_ = false;
+  bool in_active_ffi_callback_ = false;
   bool in_passive_ffi_callback_ = false;
 
   // Holds the list of focusable annot types.
